### ISSUE TITLE
cuda

### ISSUE THREAD

**Issue by smiler96** [smiler96/CUDANet#6]
```python

**Comment by smiler96** [comment]
## mddm_sig
```python
import torch
import torch.nn as nn
from Common.base import conv_
from Common.residual import RG
from MMDM.acblock import ACB_Diaglock, ACBlock
from Common.subpixel import PixelShuffle, PixelUnShuffle
import torchsummary

class FERM(nn.Module):
    def __init__(self, in_c=3, out_c=3, channels=64, n_rb=20, n_rg=10, cov_t='navie', bn=False, act='ReLU'):
        super(FERM, self).__init__()

        if cov_t == 'navie':
            self.conv = conv_
        else:
            self.conv = ACBlock

        self.in_channel = in_c
        self.out_channel = out_c
        self.n_feats = channels
        self.n_resblocks = n_rb
        self.n_resgroups = n_rg
        self.bn = bn
        self.act = act

        head = [self.conv(self.in_channel, self.n_feats)]

        body = [
            RG(self.conv, self.n_feats, bn=self.bn, act=self.act, n_resblocks=self.n_resblocks) \
            for _ in range(self.n_resgroups)]
        body.append(self.conv(self.n_feats, self.n_feats))

        tail = [
            self.conv(self.n_feats, self.n_feats),
            self.conv(self.n_feats, self.out_channel)]

        self.head = nn.Sequential(*head)
        self.body = nn.Sequential(*body)
        self.tail = nn.Sequential(*tail)

    def forward(self, x):
        x = self.head(x)

        res = self.body(x)
        res = x + res

        x = self.tail(res)

        return x

class MSFE(nn.Module):
    def __init__(self, in_c=64, out_c=64, n_features=64, n_rb=4, n_rg=3, cov_t='acb'):
        super(MSFE, self).__init__()

        self.in_channel = in_c
        self.out_channel = out_c
        self.n_features = n_features
        self.n_resblocks = n_rb
        self.n_resgroups = n_rg
        if cov_t == 'navie':
            self.conv = conv_
        else:
            self.conv = ACBlock

        self.branch1 = nn.Sequential(*[
            PixelUnShuffle(2),
            FERM(in_c=self.in_channel * 4, out_c=self.out_channel * 4, channels=self.n_features, n_rg=self.n_resgroups, n_rb=self.n_resblocks, cov_t=cov_t),
            PixelShuffle(2),
            self.conv(in_channels=self.out_channel, out_channels=self.out_channel),
            self.conv(in_channels=self.out_channel, out_channels=self.out_channel),
        ])

        self.branch2 = nn.Sequential(*[
            PixelUnShuffle(4),
            FERM(in_c=self.in_channel * 16, out_c=self.out_channel * 16, channels=self.n_features, n_rg=self.n_resgroups, n_rb=self.n_resblocks, cov_t=cov_t),
            PixelShuffle(4),
            self.conv(in_channels=self.out_channel, out_channels=self.out_channel),
            self.conv(in_channels=self.out_channel, out_channels=self.out_channel),
        ])

        self.branch3 = nn.Sequential(*[
            PixelUnShuffle(8),
            FERM(in_c=self.in_channel * 64, out_c=self.out_channel * 64, channels=self.n_features, n_rg=self.n_resgroups, n_rb=self.n_resblocks, cov_t=cov_t),
            PixelShuffle(8),
            self.conv(in_channels=self.out_channel, out_channels=self.out_channel),
            self.conv(in_channels=self.out_channel, out_channels=self.out_channel),
        ])

    def forward(self, x):
        x0 = x
        x1 = self.branch1(x)
        x2 = self.branch2(x)
        x3 = self.branch3(x)
        x = torch.cat([x0, x1, x2, x3], dim=1)
        return x

class MMDM_Sig(nn.Module):
    def __init__(self, in_c=3, out_c=3, n_feas=64, n_rb=20, n_rg=10, cov_t='navie'):
        super(MMDM_Sig, self).__init__()

        self.n_feas = n_feas
        self.n_rb = n_rb
        self.n_rg = n_rg

        self.in_c = in_c
        self.out_c = out_c

        if cov_t == 'navie':
            self.conv = conv_
        else:
            self.conv = ACBlock

        self.head = nn.Sequential(*[
            PixelUnShuffle(2),
            self.conv(in_channels=in_c * 4, out_channels=self.n_feas),
        ])

        self.ferm = FERM(in_c=self.n_feas, out_c=self.n_feas, channels=self.n_feas, n_rb=self.n_rb, n_rg=self.n_rg, cov_t=cov_t)
        self.msfe = MSFE(in_c=self.n_feas, out_c=self.n_feas, n_features=self.n_feas,  n_rb=5, n_rg=4, cov_t=cov_t)

        self.decoder = nn.Sequential(*[
            self.conv(in_channels=self.n_feas*4, out_channels=self.n_feas),
            nn.ReLU(),
            PixelShuffle(2),
            self.conv(in_channels=self.n_feas//4, out_channels=out_c),
            nn.Sigmoid(),
        ])

    def forward(self, x):
        x = self.head(x)
        _ferm = self.ferm(x)
        _msfe = self.msfe(_ferm)
        x = self.decoder(_msfe)

        return x


if __name__ == "__main__":
    import numpy as np


    model = MMDM_Sig(n_feas=32, n_rb=10, n_rg=2)
    # torchsummary.summary(model, (3, 128, 128))
    # print(torchsummary.summary(model, (3, 512, 512)))
    x = np.random.uniform(0, 1, [1, 3, 512, 512]).astype(np.float32)
    x = torch.tensor(x)

    y = model(x)
    print(y.shape)
```

**Comment by smiler96** [comment]
## acb
```python
class ACBlock(nn.Module):

    def __init__(self, in_channels, out_channels, kernel_size=3, stride=1, padding=1, dilation=1, groups=1, padding_mode='zeros', deploy=False,
                 use_affine=True):
        super(ACBlock, self).__init__()
        self.deploy = deploy
        if deploy:
            self.fused_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels, kernel_size=(kernel_size,kernel_size), stride=stride,
                                      padding=padding, dilation=dilation, groups=groups, bias=True, padding_mode=padding_mode)
        else:
            self.square_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels,
                                         kernel_size=(kernel_size, kernel_size), stride=stride,
                                         padding=padding, dilation=dilation, groups=groups, bias=False,
                                         padding_mode=padding_mode)

            center_offset_from_origin_border = padding - kernel_size // 2
            ver_pad_or_crop = (padding, center_offset_from_origin_border)
            hor_pad_or_crop = (center_offset_from_origin_border, padding)
            if center_offset_from_origin_border >= 0:
                self.ver_conv_crop_layer = nn.Identity()
                ver_conv_padding = ver_pad_or_crop
                self.hor_conv_crop_layer = nn.Identity()
                hor_conv_padding = hor_pad_or_crop
            else:
                self.ver_conv_crop_layer = CropLayer(crop_set=ver_pad_or_crop)
                ver_conv_padding = (0, 0)
                self.hor_conv_crop_layer = CropLayer(crop_set=hor_pad_or_crop)
                hor_conv_padding = (0, 0)
            self.ver_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels, kernel_size=(kernel_size, 1),
                                      stride=stride,
                                      padding=ver_conv_padding, dilation=dilation, groups=groups, bias=False,
                                      padding_mode=padding_mode)

            self.hor_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels, kernel_size=(1, kernel_size),
                                      stride=stride,
                                      padding=hor_conv_padding, dilation=dilation, groups=groups, bias=False,
                                      padding_mode=padding_mode)

    def forward(self, input):
        if self.deploy:
            return self.fused_conv(input)
        else:
            square_outputs = self.square_conv(input)
            vertical_outputs = self.ver_conv_crop_layer(input)
            vertical_outputs = self.ver_conv(vertical_outputs)
            horizontal_outputs = self.hor_conv_crop_layer(input)
            horizontal_outputs = self.hor_conv(horizontal_outputs)
            result = square_outputs + vertical_outputs + horizontal_outputs
            return result


class ACB_Diaglock(nn.Module):

    def __init__(self, in_channels, out_channels, kernel_size=3, stride=1, padding=1, dilation=1, groups=1, padding_mode='zeros'):
        super(ACB_Diaglock, self).__init__()

        self.square_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels,
                                     kernel_size=(kernel_size, kernel_size), stride=stride,
                                     padding=padding, dilation=dilation, groups=groups, bias=False,
                                     padding_mode=padding_mode)

        self.diag_l_conv = Conv_Diag_L(in_channels=in_channels, out_channels=out_channels)
        self.diag_r_conv = Conv_Diag_R(in_channels=in_channels, out_channels=out_channels)

        center_offset_from_origin_border = padding - kernel_size // 2
        ver_pad_or_crop = (padding, center_offset_from_origin_border)
        hor_pad_or_crop = (center_offset_from_origin_border, padding)
        if center_offset_from_origin_border >= 0:
            self.ver_conv_crop_layer = nn.Identity()
            ver_conv_padding = ver_pad_or_crop
            self.hor_conv_crop_layer = nn.Identity()
            hor_conv_padding = hor_pad_or_crop
        else:
            self.ver_conv_crop_layer = CropLayer(crop_set=ver_pad_or_crop)
            ver_conv_padding = (0, 0)
            self.hor_conv_crop_layer = CropLayer(crop_set=hor_pad_or_crop)
            hor_conv_padding = (0, 0)
        self.ver_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels, kernel_size=(kernel_size, 1),
                                  stride=stride,
                                  padding=ver_conv_padding, dilation=dilation, groups=groups, bias=False,
                                  padding_mode=padding_mode)

        self.hor_conv = nn.Conv2d(in_channels=in_channels, out_channels=out_channels, kernel_size=(1, kernel_size),
                                  stride=stride,
                                  padding=hor_conv_padding, dilation=dilation, groups=groups, bias=False,
                                  padding_mode=padding_mode)

    def forward(self, _input):
        square_outputs = self.square_conv(_input)
        vertical_outputs = self.ver_conv_crop_layer(_input)
        vertical_outputs = self.ver_conv(vertical_outputs)
        horizontal_outputs = self.hor_conv_crop_layer(_input)
        horizontal_outputs = self.hor_conv(horizontal_outputs)

        diag_l_outs = self.diag_l_conv(_input)
        diag_r_outs = self.diag_r_conv(_input)

        result = square_outputs + vertical_outputs + horizontal_outputs + diag_l_outs + diag_r_outs
        return result
```

**Comment by smiler96** [comment]
import torch.nn as nn
from Common.base import Conv_Diag_L, Conv_Diag_R

class CropLayer(nn.Module):
 
    def __init__(self, crop_set):
        super(CropLayer, self).__init__()
        self.rows_to_crop = - crop_set[0]
        self.cols_to_crop = - crop_set[1]
        assert self.rows_to_crop >= 0
        assert self.cols_to_crop >= 0

    def forward(self, input):
        if self.rows_to_crop == 0 and self.cols_to_crop == 0:
            return input
        elif self.rows_to_crop > 0 and self.cols_to_crop == 0:
            return input[:, :, self.rows_to_crop:-self.rows_to_crop, :]
        elif self.rows_to_crop == 0 and self.cols_to_crop > 0:
            return input[:, :, :, self.cols_to_crop:-self.cols_to_crop]
        else:
            return input[:, :, self.rows_to_crop:-self.rows_to_crop, self.cols_to_crop:-self.cols_to_crop]
